packages/contracts/contracts/votings/majority-voting/ERC20Voting.sol

```diff
@@ -0,0 +1,157 @@
+
+    mapping(uint256 => uint64) snapshotBlocks;
+
+    function getSnapshotBlock(uint256 _voteId) public view returns (uint64) {
+        return snapshotBlocks[_voteId];
+    }
+
+    /// @dev describes the version and contract for GSN compatibility.
+    function versionRecipient() external view virtual override returns (string memory) {
+        return "0.0.1+opengsn.recipient.ERC20Voting";
+    }
+
+    /// @dev Used for UUPS upgradability pattern
+    /// @param _dao The DAO contract of the current DAO
+    function initialize(
+        IDAO _dao,
+        address _gsnForwarder,
+        uint64 _participationRequiredPct,
+        uint64 _supportRequiredPct,
+        uint64 _minDuration,
+        ERC20VotesUpgradeable _token
+    ) public initializer {
+        __MajorityVotingBase_init(
+            _dao,
+            _gsnForwarder,
+            _participationRequiredPct,
+            _supportRequiredPct,
+            _minDuration
+        );
+
+        token = _token;
+    }
+
+    /// @notice Create a new vote on this concrete implementation
+    /// @param _proposalMetadata The IPFS hash pointing to the proposal metadata
+    /// @param _actions the actions that will be executed after vote passes
+    /// @param _startDate state date of the vote. If 0, uses current timestamp
+    /// @param _endDate end date of the vote. If 0, uses _start + minDuration
+    /// @param _executeIfDecided Configuration to enable automatic execution on the last required vote
+    /// @param _choice Vote choice to cast on creation
+    function newVote(
+        bytes calldata _proposalMetadata,
+        IDAO.Action[] calldata _actions,
+        uint64 _startDate,
+        uint64 _endDate,
+        bool _executeIfDecided,
+        VoterState _choice
+    ) external override returns (uint256 voteId) {
+        uint64 snapshotBlock = getBlockNumber64() - 1;
+
+        uint256 votingPower = token.getPastTotalSupply(snapshotBlock);
+        if (votingPower == 0) revert VotePowerZero();
+        
+        voteId = votesLength++;
+
+        // calculate start and end time for the vote
+        uint64 currentTimestamp = getTimestamp64();
+
+        if (_startDate == 0) _startDate = currentTimestamp;
+        if (_endDate == 0) _endDate = _startDate + minDuration;
+
+        if(_endDate - _startDate <  minDuration || _startDate < currentTimestamp)
+            revert VoteTimesForbidden({
+                current: currentTimestamp,
+                start: _startDate,
+                end: _endDate,
+                minDuration: minDuration
+            });
+
+        // create a vote.
+        snapshotBlocks[voteId] = snapshotBlock;
+
+        Vote storage vote_ = votes[voteId];
+        vote_.startDate = _startDate;
+        vote_.endDate = _endDate;
+        vote_.supportRequiredPct = supportRequiredPct;
+        vote_.participationRequiredPct = participationRequiredPct;
+        vote_.votingPower = votingPower;
+
+        for (uint256 i; i < _actions.length; i++) {
```

Comment:
don't want to annoy but if we anyways touch this could we directly put it into the ``unchecked`` closure and store the actions length on the stack to save some gas.

Reply:
Done